[WebGPUSwift] re-implement copyTextureToTexture in Swift

nmahendru opened the pull request:
#### 4309ab042c6cdca3ed5e0cfe9230715b1ff62f5e
<pre>
[WebGPUSwift] re-implement copyTextureToTexture in Swift
<a href="https://bugs.webkit.org/show_bug.cgi?id=284044">https://bugs.webkit.org/show_bug.cgi?id=284044</a>
<a href="https://rdar.apple.com/140917930">rdar://140917930</a>

Reviewed by Mike Wyrzykowski.

It&apos;s a 1-for-1 replacement.
Tested at desk with:

./LayoutTests/http/tests/webgpu/webgpu/api/operation/command_buffer/copyTextureToTexture.html &lt;== Timeout(Same behavior as the non Swift implementation)
./LayoutTests/http/tests/webgpu/webgpu/api/validation/encoding/cmds/copyTextureToTexture.html &lt;== PASS

Also, after discussing with Mike, adding `[ Release ] http/tests/webgpu/webgpu/api/validation/encoding/cmds/copyTextureToTexture.html` to mac-wk2
as that test is actually working for non swift builds as well.

* Source/WTF/wtf/CMakeLists.txt:
* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::errorValidatingCopyTextureToTexture const):
(WebGPU::CommandEncoder::copyTextureToTexture):
(WebGPU::errorValidatingCopyTextureToTexture): Deleted.
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(CommandEncoder_copyTextureToTexture_thunk(commandEncoder:source:destination:copySize:)):
(WebGPU.copyTextureToTexture(_:destination:copySize:)):
* Source/WebGPU/WebGPU/Queue.h:
* Source/WebGPU/WebGPU/Queue.swift:
(WebGPU.writeBuffer(_:bufferOffset:data:)):

Canonical link: <a href="https://commits.webkit.org/287423@main">https://commits.webkit.org/287423@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f35bc95c9b70e22c79eb599ebd9aa4648b56f66b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79638 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58638 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33033 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84182 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30684 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67732 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6925 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62253 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20107 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82704 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52313 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72535 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42561 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49656 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26687 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29110 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72679 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70795 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27145 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85595 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/78768 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6890 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4801 "Found 1 new test failure: http/wpt/mediarecorder/MediaRecorder-multiple-start-stop.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70501 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7060 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68380 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69746 "Found 6 new API test failures: /TestWebKit:WebKit2TextFieldBeginAndEditEditingTest.TextFieldDidBeginAndEndEditingEventsInReadOnlyField, /TestWebKit:WebKit2.ProvisionalURLAfterWillSendRequestCallback, /TestWebKit:WebKit.HitTestResultNodeHandle, /TestWebKit:WebKit2TextFieldBeginAndEditEditingTest.TextFieldDidBeginShouldNotBeDispatchedForAlreadyFocusedField, /TestWebKit:WebKit2UserMessageRoundTripTest.WKURLRequestRef, /TestWebKit:WebKit.ParentFrame (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13764 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12675 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/101108 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12300 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6839 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24673 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6730 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10224 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8527 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (commented):
Can we add:
```
[ Release ] http/tests/webgpu/webgpu/api/operation/command_buffer/copyTextureToTexture.html
[ Release ] http/tests/webgpu/webgpu/api/validation/encoding/cmds/copyTextureToTexture.html
```
to `LayoutTests/platform/mac-wk2/TestExpectations` if they are passing?

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -1731,6 +1732,7 @@ static bool areCopyCompatible(WGPUTextureFormat format1, WGPUTextureFormat forma
     return nil;
 }
 
+#if ! ENABLE(WEBGPU_SWIFT)
```
I think it's fine but the space between the `!` and `ENABLE` seems unintentional perhaps?

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -52,6 +52,11 @@ public func CommandEncoder_copyTextureToBuffer_thunk(commandEncoder: WebGPU.Comm
 }
 
+@_expose(Cxx)
+public func CommandEncoder_copyTextureToTexture_thunk(commandEncoder:WebGPU.CommandEncoder, source: WGPUImageCopyTexture, destination: WGPUImageCopyTexture, copySize: WGPUExtent3D) {
```
missing space after commandEncoder:

nmahendru replied:
fixed

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -743,4 +748,186 @@ extension WebGPU.CommandEncoder {
+
+        // https://gpuweb.github.io/gpuweb/#dom-gpucommandencoder-copytexturetotexture
+
+        guard  self.prepareTheEncoderState() else  {
```
extra space after guard

rr-codes replied:
and after else

nmahendru replied:
fixed

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -743,4 +748,186 @@ extension WebGPU.CommandEncoder {
+            return
+        }
+        let error = self.errorValidatingCopyTextureToTexture(source, destination, copySize)
+        guard error != nil else {
```
shouldn't this be ==

nmahendru replied:
It should be.. I haven't even self reviewed this yet...or tested this yet.. I just pushed to try on a different machine.. I am getting a linker error at the moment.. 
Yes I probably should have made this draft.. my bad..

nmahendru committed:
[WebGPUSwift] re-implement copyTextureToTexture in Swift
https://bugs.webkit.org/show_bug.cgi?id=284044
rdar://140917930

Reviewed by Mike Wyrzykowski.

It's a 1-for-1 replacement.
Tested at desk with:

./LayoutTests/http/tests/webgpu/webgpu/api/operation/command_buffer/copyTextureToTexture.html <== Timeout(Same behavior as the non Swift implementation)
./LayoutTests/http/tests/webgpu/webgpu/api/validation/encoding/cmds/copyTextureToTexture.html <== PASS

Also, after discussing with Mike, adding `[ Release ] http/tests/webgpu/webgpu/api/validation/encoding/cmds/copyTextureToTexture.html` to mac-wk2
as that test is actually working for non swift builds as well.

* Source/WTF/wtf/CMakeLists.txt:
* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::errorValidatingCopyTextureToTexture const):
(WebGPU::CommandEncoder::copyTextureToTexture):
(WebGPU::errorValidatingCopyTextureToTexture): Deleted.
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(CommandEncoder_copyTextureToTexture_thunk(commandEncoder:source:destination:copySize:)):
(WebGPU.copyTextureToTexture(_:destination:copySize:)):
* Source/WebGPU/WebGPU/Queue.h:
* Source/WebGPU/WebGPU/Queue.swift:
(WebGPU.writeBuffer(_:bufferOffset:data:)):

Canonical link: https://commits.webkit.org/287423@main
